### DISCUSSION TITLE
Add a new :complete-function to capf to support lsp completion better

### DISCUSSION THREAD

**Discussion by zbelial** [company-mode/company-mode#1422] [Ideas]
I wish to improve user experience when using company-capf with lsp clients.  Here is my proposal.  I'll describe briefly how company + lsp capf works  and then explain how `:complete-function` can help here.

When you press ENTER after selecting a candidate from the popup, company + lsp clients' capf work in this way:
1. company deletes the prefix (if any) input already
2. company inserts candidate's content
3. company will call lsp capf's `:exit-function`, in which it deletes what has been inserted in step 2, and then inserts what's really desired (may need to expand a snippet via yasnippet)

here is an example. 
![图片](https://github.com/company-mode/company-mode/assets/6436456/e27c68af-ef54-4547-be85-45f9f95a5579)

I have input "add", this is the prefix.  Now I'll simulate step by step how company + lsp capf works
1. company deletes "add"
2. company inserts "addRound(BigDecimal a, BigDecimal b, Integer scale) : BigDecimal"
3. company deletes "addRound(BigDecimal a, BigDecimal b, Integer scale) : BigDecimal" and then inserts "addRound(a, b, scale)", 
Now you'll get what you really want (a snippet pre-filled with what jdt.ls thinks suitable as arguments)

![图片](https://github.com/company-mode/company-mode/assets/6436456/6cc8e4ea-152e-4a01-bfb7-0bc7ae0ae1d6)

Now I'll describe how `:complete-function` can improve this situation. With `:complete-function`, company + lsp capf works in this way:
1. company calls `:complete-function`, in which it deletes the prefix, and then inserts what's desired. (Very similar to what `:exit-function` does in this lsp case)

Because of no need to insert "addRound(BigDecimal a, BigDecimal b, Integer scale) : BigDecimal" and then delete it, and then insert what needed, compeltion will be more smooth and with less filicker. 

What do you think?

**Comment by dgutov** [comment]
Good topic, and it's something that was mentioned before. Let me turn it into an issue.
